sycl/source/detail/device_impl.cpp

```diff
@@ -603,29 +603,50 @@ bool device_impl::has(aspect Aspect) const {
     return Result != nullptr;
   }
   case aspect::ext_oneapi_graph: {
-    size_t ResultSize = 0;
-    bool CallSuccessful = getPlugin()->call_nocheck<PiApiKind::piDeviceGetInfo>(
-                              MDevice, PI_DEVICE_INFO_EXTENSIONS, 0, nullptr,
-                              &ResultSize) == PI_SUCCESS;
-    if (!CallSuccessful || ResultSize == 0) {
+    pi_bool SupportsCommandBuffers = false;
+    bool CallSuccessful =
+        getPlugin()->call_nocheck<PiApiKind::piDeviceGetInfo>(
+            MDevice, PI_EXT_ONEAPI_DEVICE_INFO_COMMAND_BUFFER_SUPPORT,
+            sizeof(SupportsCommandBuffers), &SupportsCommandBuffers,
+            nullptr) == PI_SUCCESS;
+    if (!CallSuccessful) {
       return PI_FALSE;
     }
 
-    std::unique_ptr<char[]> Result(new char[ResultSize]);
-    CallSuccessful = getPlugin()->call_nocheck<PiApiKind::piDeviceGetInfo>(
-                         MDevice, PI_DEVICE_INFO_EXTENSIONS, ResultSize,
-                         Result.get(), nullptr) == PI_SUCCESS;
+    pi_bool SupportsCommandBufferUpdate = false;
+    CallSuccessful =
+        getPlugin()->call_nocheck<PiApiKind::piDeviceGetInfo>(
+            MDevice, PI_EXT_ONEAPI_DEVICE_INFO_COMMAND_BUFFER_UPDATE_SUPPORT,
+            sizeof(SupportsCommandBufferUpdate), &SupportsCommandBufferUpdate,
+            nullptr) == PI_SUCCESS;
+    if (!CallSuccessful) {
+      return PI_FALSE;
+    }
 
+    return SupportsCommandBuffers && SupportsCommandBufferUpdate;
+  }
+  case aspect::ext_oneapi_limited_graph: {
```

Comment:
This new aspect should be documented somewhere. If I'm reading it correctly, this and `ext_oneapi_graph` are mutually exclusive. I think a better design is to have `ext_oneapi_graph` mean support for basic features and then have a new `ext_oneapi_graph_arg_update` or similar, which is true if graphs supports this new argument updating. Since `ext_oneapi_graph_arg_update` would only be true if `ext_oneapi_graph` is true, users could either check both or rely on update support when needed. @gmlueck - Thoughts?

Reply:
Yes this aspect, and these new APIs in general, are documented in a separate PR since the changes are large: https://github.com/intel/llvm/pull/12486

I agree they do not need to be mutually exclusive, that is just how I interpreted the request. I do think the naming and description of `ext_oneapi_limited_graph` becomes a bit confusing if they are not mutually exclusive, but that is probably not a big deal.

Reply:
I like the approach that @gmlueck is suggesting.
My understanding was that we add a new aspect `ext_oneapi_graph` that is true when graph is fully supported.
And I think that a superset makes sense here.

How about we are a little bit more specific and name the subset `ext_oneapi_graph_immutable` instead of `ext_oneapi_limited_graph`?

Reply:
> How about we are a little bit more specific and name the subset `ext_oneapi_graph_immutable` instead of `ext_oneapi_limited_graph`?

That's OK with me.  Do we feel confident that the "limited" devices will support all graph features aside from mutating the kernel arguments, etc.?  The advantage of the name `ext_oneapi_limited_graph` is that it can be a catch-all for all features that old devices don't support.

Reply:
I think it's pretty likely that other features we're looking to add soon won't be supported by all devices. For example, graph-owned memory allocations won't be available in an OpenCL command-buffer backend.  External event API for synchronizing in/out of the graph likewise won't be available in an OpenCL backend. The node level profiling API we've got a prototype implementation of is only available on Level Zero right now.

So that's limitations in non-L0 devices, but in terms of L0 specifically the L0 team has powerpoint level ideas of features related to graphs which aren't mutability related, which if they delivered wouldn't be available on all devices but also wouldn't be covered by a mutability specific limitation aspect.

So overall i have a preference for the more generic name for the aspect, rather than focusing on the mutability limitation.

Reply:
Let's use `ext_oneapi_limited_graph` for now.  We can re-evaluate as we add more features and see what support matrix we have.  This is an experimental feature, so we can change the names of the aspects later if necessary.

Reply:
I've update this so that the aspects are no longer mutually exclusive.